compiler/src/yul/selectors.rs

```diff
@@ -0,0 +1,175 @@
+use crate::errors::CompileError;
+use crate::yul::base;
+use tiny_keccak::{Hasher, Keccak};
+use yultsur::yul;
+
+/// Builds a switch statement from the contract ABI.
+/// The switch's expression is the 4 left-most bytes in the calldata and each case is
+/// defined as the keccak value of each function's signature (without return data).
+pub fn switch(functions: Vec<&ethabi::Function>) -> Result<yul::Switch, CompileError> {
+    let cases = functions
+        .into_iter()
+        .map(|function| case(function))
+        .collect::<Result<Vec<yul::Case>, CompileError>>()?;
+
+    Ok(yul::Switch {
+        expression: expression(),
+        cases,
+    })
+}
+
+/// Builds an expression that loads the first 4 bytes of the calldata.
+pub fn expression() -> yul::Expression {
+    yul::Expression::FunctionCall(yul::FunctionCall {
+        identifier: base::untyped_identifier("shr"),
+        arguments: vec![
+            base::untyped_literal_expr("224"),
+            yul::Expression::FunctionCall(yul::FunctionCall {
+                identifier: base::untyped_identifier("calldataload"),
+                arguments: vec![base::untyped_literal_expr("0")],
+            }),
+        ],
+    })
+}
+
+/// Builds a switch case from the function. It matches the selector and calls
+/// the function as described in the ABI. The value (if any) returned by the
+/// function is stored in memory and returned by the contract.
+///
+/// Currently, this assumes each input and the single output is 256 bits.
+/// TODO: Handle types of different sizes: https://solidity.readthedocs.io/en/v0.6.2/abi-spec.html#types
+pub fn case(function: &ethabi::Function) -> Result<yul::Case, CompileError> {
+    let selector = Some(selector_literal(function.signature()));
+
+    if function.outputs.is_empty() {
+        return Ok(yul::Case {
+            literal: selector,
+            block: yul::Block {
+                statements: vec![yul::Statement::Expression(yul::Expression::FunctionCall(
+                    yul::FunctionCall {
+                        identifier: base::untyped_identifier("foo"),
```

Comment:
might not need to update here if this is temporary but i expect we can grab some `identifier` from the `ethabi::Function`

Reply:
Oh, woops. This got fixed [later on](https://github.com/g-r-a-n-t/rust-vyper/blob/types/compiler/src/yul/runtime/abi.rs#L25), probably after trying to run a function not named "foo".